### DISCUSSION TITLE
[RFC] POCL_DEVICES: driver toggle vs number of devices

### DISCUSSION THREAD

**Discussion by Oblomov** [pocl/pocl#1389] [Ideas]
Some drivers (VULKAN, hsa) currently use the `POCL_DEVICES` as a mere toggle: they are disabled by default, but will probe for devices if present in the string, and return all the probed devices. Other drivers (basic, pthread, CUDA) use the number of occurrences to determine how many devices to probe for. This is a bit inconsistent and somewhat confusing, especially since the CPU drivers use the number to “clone” the (single) device they expose, while CUDA for example uses it to _clamp_ the number of devices, and will in fact throw an error if the requested number of devices is higher than the number of probed devices.

While I understand the rationale for the CPU devices to use `POCL_DEVICES` as a “device count”, I'm perplexed about the usefulness of this use of the feature in drivers that naturally support multiple devices. It's especially frustrating when you want to enable some normally-disabled device: e.g., if I want to enable VULKAN and keep the CUDA devices, then I have to set `POCL_DEVICES` to `VULKAN CUDA CUDA CUDA CUDA` etc with as may CUDA devices as my machine has (think e.g. about a large server with 8 GPUs).

One argument that could be made in favor of treating it as a count even for devices such as CUDA is that this would allow some degree of control on which devices are exposed. However, the actual usefulness of this is debatable, since there's no guarantee about device probe order, and when the devices are not all copies of each other (think e.g. of a small display-only NVIDIA GPU + a workstation GPU used only for compute), or in drivers such as Vulkan or HSA where the multiple devices can be completely different from each other, simply limiting by count is generally not what the user would want.

My recommendation would be for most drivers to treat POCL_DEVICES simply as a boolean (keeping the exception only for the CPU), and introduce some other mechanism (either a different environment variable or an extended syntax of the same) to handle device filtering for multi-device drivers.

What does the pocl team think of this?

**Comment by pjaaskel** [comment]
The original idea was indeed to be able to define/restrict the platform's devices explicitly with the POCL_DEVICES list. It has worked well for testing with the CPU devices (and ttasim), but I see it's just "unimplemented" for CUDA and Vulkan where it's more of a toggle. 

To improve this original idea of a nice way to setup the platform explicitly, there could be some sort of syntax for the probing. E.g. 'vulkan*' would add all found Vulkan devices, "vulkan vulkan" would add (at most) two and "vulkan16" at most 16. There might be thousands of devices in a PoCL platform in the future after we contribute the PoCL-R code that can drive distributed OpenCL devices.

There is still a lack for a full probe env (could be POCL_DEVICES=*) which rely solely on probing to get the devices added to the platform.

**Comment by Oblomov** [comment]
For things like vulkan and cuda (and hsa?), I don't think number-based restrictions are a good idea, though. Some kind of filtering based on device properties (name, UUID, PCI address, etc) would probably be better.

Maybe the syntax could be extended to something like `driver/devicespec`, with `*` meaning “any”, and with with the possibility to omit the `/devicespec` part, implying `/*`. The actual meaning of the `devicespec`, aside from `*` meaning “any”, could be driver-specific, and the same for the behavior of repeated matches. So for example:

* `driver` or equivalently `driver/*` would mean “any device probed by `driver`”;
* `*` would mean “any device probed by any driver”;
* `*/nvidia` would mean “any device probed by any driver, that matches the device-spec `nvidia`”;

This would be a breaking change, but it could probably be done in a mostly-backwards compatible way (with the possible exception of the behavior of the `cuda` driver).

**Comment by pjaaskel** [comment]
SGTM!

**Comment by Oblomov** [comment]
I can work on a draft implementation of this idea, but the way I see this happening at least for the drivers I'm more familiar with would require the filtering to happen at init rather than probe time, meaning that a filtered driver will effectively init less devices than claimed during probe. Is this a potential issue? Should init return a special value in this case? The alternative would be to double-query device properties during probe (for filtering) and init, but I can't say I would be too happy with this solution.

**Comment by franz** [comment]
> Is this a potential issue?

... possibly. Someone has to go through the code and figure that out. At minimum, all functions related to context creation need to be aware of these uninitialized device structs. This could be enough, but i'm not 100% sure. 

> I can't say I would be too happy with this solution.

I can't say i'm too happy with either of them... maybe this points to a problem with the driver API: having separate probe & init. The only real usefulness of their separation, is that we know the size of `pocl_devices` array before calling each device's init, but this can be solved fairly simply. So IMO you can go ahead and merge probe+init if it makes your life easier.
